# <font color=blue>Ceros de Funciones en una variable</font>

**Problema**: 
>Dada una función continua $ \ f:[a,b]\subset\mathbb R\to\mathbb R$, 

>hallar <font color=red>$x_*$</font>$\in(a,b)$ tal que 
 $f$(<font color=red>$x_*$</font>)$=0$

## Bisección

- requiere un intervalo con la raíz
- divide el intervalo a la mitad recursivamente
- elige subintervalo donde la función cambia de signo
- cero $\approx$ punto medio de subintervalo

**Ejemplo 1:**

Calcular aproximación $x_n$ del cero del polinomio 

> $f(x)=3.7x^8-7.5x^3+3x-10.3:\mathbb R\to\mathbb R$

en el intervalo $[-1.5,-0.5]$ 

Criterio de paro 
> $|f(x_n)| \leq 0.05$


In [14]:
using Plots
plotly()
fun1(x::Union{Float64,Array{Float64,1}}) = 3.7x.^8 -7.5x.^3 + 3.0x - 10.3 

fun1 (generic function with 1 method)

In [3]:
it = 0

0

In [16]:
println("Dame a:")
sleep(1)
a = parse(Float64,readline())
println("Dame b:")
sleep(1)
b = parse(Float64,readline())
m  = 0.5*(a+b)
fa = fun1(a) 
fb = fun1(b)
fm = fun1(m)
pt = @sprintf "f(a)       =%1.3e\nf((a+b)/2) =%1.3e\nf(b)       =%1.3e" fa fm fb 
Iam = @sprintf "nuevo intervalo = [ %1.3e , %1.3e ]" a m
Imb = @sprintf "nuevo intervalo = [ %1.3e , %1.3e ]" m b
println("iteración = $it")
println("(a+b)/2 = $m")
println(pt)
(fa*fm <0) ? (println(Iam)) : (println(Imb))
it += 1;

Dame a:
STDIN> -1.046875
Dame b:
STDIN> -1.03125
iteración = 6
(a+b)/2 = -1.0390625
f(a)       =5.020e-01
f((a+b)/2) =2.380e-02
f(b)       =-4.357e-01
nuevo intervalo = [ -1.039e+00 , -1.031e+00 ]


In [17]:
xpa = collect(linspace(a,b,101))
ypa = fun1(xpa)
plot(xpa,ypa,color="blue",label="f(x)=3.7x⁸-7.5x³+3x-10.3")
plot!([a,b],zeros(2),color="purple",label="")
scatter!([a,m,b],[fa,fm,fb],color="green",label="")

$\begin{matrix} \\ \\ \\ \\ \\  \\ \end{matrix}$

*Resultados:*
- 6 iteraciones
- aprox $\approx -1.046875$
- f(aprox) $\approx 2.380\times 10^{-2}$

<font color=red>¿Justificación?</font>

***Teorema de Valor Intermedio*** 

Sea $ \ f:[a,b]\subset\mathbb R\to\mathbb R$
una función continua tal que $f(a)\neq f(b)$. 

Entonces, para cada $v$ entre $f(a)$ y $f(b)$
existe al menos un $r\in(a,b)$ tal que 
> $f(r)=v$

**Condición para cambio de signo:**
> $f(a)f(b)<0$

Siguiente paso:
- usar ciclo para realizar biseccción
- salir del ciclo por criterio de paro
- en cada paso checa $ \ f(a)\cdot f\left(\dfrac{a+b}{2}\right) \ $ 
  para actualizar $a$ y $b$

In [16]:
a   = -1.5
b   =  -0.5
m   = 0.5*(a+b)
fm  = fun1(m)
it  = 0
xpa = collect(linspace(a,b,101))
ypa = fun1(xpa)
while ( abs(fm) > 5e-2 )
     it +=1
     fa = fun1(a)
     fb = fun1(b)
     # gráfica
     IJulia.clear_output(true)
     plot(xpa,ypa,color="blue")
     plot!([xpa[1],xpa[end]],zeros(2),color="purple",label="")
     display(scatter!([a,m,b],[fa,fm,fb],color="green",label=""))
     # iteración principal
     (fa*fm <0) ? (b=m) : (a=m)
     m = 0.5*(a+b)
     fm = fun1(m)
     println("iteraciones: $it\naprox: $m")
     sleep(2)
end

iteraciones: 6
aprox: -1.0390625


**Ejemplo 2:**

Hallar aproximación $x_n$ del cero <font color=red>$x_*$</font> de la función

> $f(x)=\sqrt{x}-\cos(x)$

en el intervalo $[0,1]$ tal que

> $|f(x_n)|< 10^{-3}$

In [11]:
using Plots; plotly(); font(19)

Plots.Font("sans-serif", 19, :hcenter, :vcenter, 0.0, RGB{N0f8}(0.0,0.0,0.0))

In [9]:
fun2(x::Union{Float64,Array{Float64,1}}) = sqrt.(x) -cos.(x) 

fun2 (generic function with 1 method)

In [13]:
pa = collect(linspace(0,1,101)) 
plot(pa,fun2(pa),color="blue",linewidth=5,grid=false,label="f(x)=√x-cos(x)",legend=:top,xlabel="x")
plot!([pa[1],pa[end]],zeros(2),color="purple",label="")

In [34]:
println("intervalo inicial [a,b]\nDame a:")
sleep(1)
a = parse(Float64,readline())
println("Dame b:")
sleep(1)
b = parse(Float64,readline())
println("Dame toleracia:")
sleep(1)
tol = parse(Float64,readline())
m   = 0.5*(a+b)
fm  = fun2(m)
it  = 0
xpa = collect(linspace(a,b,101))
ypa = fun2(xpa)
while ( abs(fm) > tol )
     it +=1
     fa = fun2(a)
     fb = fun2(b)
     # gráfica
     I = collect(linspace(a,b,101))
     IJulia.clear_output(true)
     plot(xpa,ypa,linewidth=2,color="blue",grid=false,label="f(x)=√x-cos(x)",
          legend=:top,xlabel="x")
     plot!(I,fun2(I),fill=(0,"#99c0c3"),linewidth=0,label="")
     plot!([xpa[1],xpa[end]],zeros(2),color="purple",label="")
     display(scatter!([a,m,b],[fa,fm,fb],color="green",label=""))
     # iteración principal
     ( sign(fa)*sign(fm)  <0 ) ? (b=m) : (a=m)
     m = 0.5*(a+b)
     fm = fun2(m)
     println("iteración: $it")
     pt = @sprintf "a = %1.4e f(a) = %1.4e\nm = %1.4e f(m) = %1.4e" a fun2(a) m fun2(m)
     println(pt)
     pt = @sprintf "b = %1.4e f(b) = %1.4e" b fun2(b)
     println(pt)
     println("intervalo [ $a , $b ] \naprox: $m")
     sleep(5)
end
println("ya salí")

iteración: 9
a = 6.4063e-01 f(a) = -1.3318e-03
m = 6.4160e-01 f(m) = -1.3793e-04
b = 6.4258e-01 f(b) = 1.0563e-03
intervalo [ 0.640625 , 0.642578125 ] 
aprox: 0.6416015625
ya salí
